### ISSUE TITLE
test.php not finding sockets

### ISSUE THREAD

**Issue by mlehtinen** [fernadosilva/orvfms#4]
Hi,

Any idea what could be the reason for test.php not passing? When I run _php-cgi test.php_ on Raspberry Pi / Jessie, I get:

```
PHP Warning:  fopen(/tmp/orvfms.dat): failed to open stream: No such file or directory in /home/mika/dev/orvfms/lib/orvfms/orvfms.php on line 1108
readDataFile: can't open file

Giving up searching for sockets
Content-type: text/html; charset=UTF-8





<h2>No sockets found</h2>

 Please check if all sockets are on-line and assure that they
 are not locked:
 (check WiWo app -> select socket -> more -> advanced).<p>

 In this version, locked or password protected devices are not supported.

<p>
```

I have the broadcast address set correctly, see traffic with tcpdump between the Pi and S20. Socket is not locked and can be controlled on Windows using _s20-gui.exe_ by Andrius Štikonas.

Any clues would be appreciated.

--Mika

**Comment by pelayon** [comment]
Exactly the same error in the same device. Any help is appreciated.

**Comment by fernadosilva** [comment]
Hello Mika and pelayon, 

Thank you for your reports.

The message  in the log file 

fopen(/tmp/orvfms.dat): failed to open stream: 
No such file or directory in /.../orvfms.php on line 1108
readDataFile: can't open file

is perfectly normal and it appears the first time the program is run (or each time this temporary file is removed, as result of a reboot or other). In fact, in order to avoid a lengthy discovery process, I store in /tmp/orvfms discovered sockets and some associated info. This speeds up socket access the next time. The first time the program is run this file does not exists and the program reports this error. However, the program enters automatically in discovery mode, initialisation may take one or two seconds more, but all required information is retrieved from sockets, this file is created and the program behaviour should be exactly the same, except for a slightly slower start.

What is a little bit more challenging is the <h2>No sockets found</h2> message. This means that we are sending a discovery messages to the LAN, and no reply is received. I have several people successfully using the software and the only time I got a similar report was when the sockets were locked. Anyway, please try one of these:
1. Please double check the  broadcast address (sorry for my insistence :-()!
2. Please check if there are any permissions that are missing (but any permission problem should be reported in the log file, therefore I doubt this). But the web server (in linux, user is usually www) must be able to create  sockets. 
3. In lib/orvfms/globals.php, please increase the timeout in the line define("TIMEOUT",0.3). Try 0.5 instead of 0.3 or even 1.0, just to be sure. The process will be slower, but should work if this is a timeout issue. 
4. If none of the above solve the problem, try to change line 668 of orvfms.php: where is
   
   if(++$loopCount > 3){

replace with

 if(++$loopCount > 5){
1. Check if any error or warning appear in the web server log file.

Anyway, I must say that I am not too convinced of any of the above. These constants were carefully tuned and seem to work almost always. Possibly there is some other subtle point here that I am missing.

Regards,

-- Fernando

**Comment by fernadosilva** [comment]
Hello,

I just noticed that I am running 7.0 (wheezy) and not Jessie. I would not expect any major difference, since there is even a lot of people running this code on plain Windows and MacOS. I do not exclude any minor bug in my code that turned to be more evident with an upgraded kernel. pelayon, can you please confirm your distribution? I have a reserve PI with me and I may consider to upgrade it to Jessie just to try to debug this issue.

Thank you both.

**Comment by pelayon** [comment]
The version is Jessie. I'll try probe your indacations this afternoon. 
pi@raspberrypi:~ $ cat /etc/os-release
PRETTY_NAME="Raspbian GNU/Linux 8 (jessie)"
NAME="Raspbian GNU/Linux"
VERSION_ID="8"
VERSION="8 (jessie)"

**Comment by pelayon** [comment]
Hi Fernando, 
First of all, I would like to thank you for your quick support. I tried all your instructions (broadcast address, permission, change code) but still show the same error. The PHP page runs correctly in apache in windows but in my rapberry shows this error. I would like run in may raspberry because of power consumption and manage the operation from the net (official application works really wrong and does not allow settings timers). I've tried other php codes and work, but not yours (which is what I like because It allows you to configure almost everything easily). For example, in https://www.domoticz.com/forum/viewtopic.php?f=11&t=8212#p56838 I found code that recognize the socket status amd change the state and work well in my raspberry.

Thank you Fernando, greetings from Spain

**Comment by mlehtinen** [comment]
Hi Fernando,

Problem solved! Nothing wrong with your fine code. Turns out it is my firewall rules...

After looking at the network traffic I realized that because the conversation starts with UDP broadcast for socket discovery, a stateful rule for the packets returning from the from the **unicast** IP address of the S20 socket back to the Pi is never automatically created (on the Pi side).

Experimeted by opening a "pinhole" like this and behold, the test.php passes.
`echo "" | nc -w 1 -u -4 -p 10000 192.168.1.108 10000`

So time to tune my iptables rules.

Thanks for your quick responses, your work on orvfms and sorry that I bothered you with this issue that I had caused myself;)

--Mika

**Comment by pelayon** [comment]
Hi Fernando, 

I solved the problem too. In my case my Network is 172.16.0.0/24. I modified globals.php with broadcast ip 172.16.0.255. I use wireshark to view the traffic. Three arp broadcast petitions searching for the mac of the ip 172.16.0.255 and stopped. So I change the broadcast ip in globals.php for the ip of my orvibo (the DHCP server assign always the same IP by MAC)  and run correctly.

**Comment by fernadosilva** [comment]
Hello Mika,

Glad you sorted it out! Thanks for your report, I will include a line in the README with this issue.

@pelayon : can you please check if you are suffering from some similar firewall issue? Can you try

$ sudo iptables -F INPUT

 and check if the code runs?

Thank you both 

Fernando

**Comment by fernadosilva** [comment]
@pelayon: great! Crossed comments by 2 minutes!

Glad that you were also able to sort it out. Thank you for your interest and reports.

Rgds,

Fernando

**Comment by fernadosilva** [comment]
@pelayon : reading again your post, I must say that your workaround is not convenient, and it will not work at all if you add more S20 to your network. It seems that there is something wrong on your netmask and network definition on some network device which is interpreting the broadcast address as a single IP. Can you check if the netmask in all devices is 255.255.255.0? can you please try as broadcast address 172.16.255.255?

Rgds

Fernando

**Comment by pelayon** [comment]
I found the real error in my case. Thinking about the arp broadcast petition clearly the error was in my Network... and eureka!!!. Yesterday I had update my orvibo and I had configurated my router with ip 44.121.111.1/8 (the official app works very bad and their support recommend me update). Then, I change back the IP but forgot change the netmask too. Sorry for all, the issue was caused myself,  too. The code works awsome.
